File: app-server/src/main/java/com/agent/timo/chat/ChatDelta.java
package com.agent.timo.chat;

/**
 * M1-6 SSE 下行事件协议（ChatDelta）。逐条对齐源项目 {@code apps/web/src/lib/events.ts} 的
 * ChatDelta 判别联合——前端 {@code lib/sse.ts} 只需 {@code JSON.parse(data)} 后按 {@code type}
 * 分发，{@code store/chat-delta.ts} 的 reducer 语义零改动（spec §3.3）。
 *
 * <p>SSE 帧形如 {@code id:<seq>\nevent:<type>\ndata:{"type":"...",...}}：{@code type} 同时出现在
 * event 名与 data 里（自描述，前端按任一分发均可）。所有事件由 {@link SessionEventBuffer} 分配
 * 会话内单调 seq，{@code Last-Event-Id} 断线续传按 seq 补发。
 *
 * <p>M1-6 投影子集：源有而 AgentScope 2.0.2 无对应语义的事件（compaction_* / tree /
 * fork_messages / session_title / runtime_invalidated / switch_response / ready）不定义，
 * 待 M1-11/M2 引入对应能力时补充。messages_loaded 在 M1-6 恒为空数组（无持久化，重连状态
 * 依赖 seq 缓冲补发；M1-11 落 DB 后回真实历史）。
 */
public sealed interface ChatDelta {

    /** 判别字段，同时用作 SSE 的 {@code event:} 名（各 record 的首个组件天然实现之）。 */
    String type();

    /** 会话元信息（SSE 连接建立 / 全量重建时首发）。 */
    record SessionInfo(String type, String sessionId, String sessionName) implements ChatDelta {
        public static SessionInfo of(String sessionId, String sessionName) {
            return new SessionInfo("session_info", sessionId, sessionName);
        }
    }

    /**
     * 历史消息全量（仅全量重建时发）。载荷是源 pi {@code AgentMessage} 形状的数组
     * （{@code session_entries.payload} 原文，见 SessionEntryProjector 契约注释）；
     * M1-6 曾恒为空数组。
     */
    record MessagesLoaded(String type, java.util.List<Object> messages) implements ChatDelta {
        public static MessagesLoaded empty() {
            return new MessagesLoaded("messages_loaded", java.util.List.of());
        }

        public static MessagesLoaded of(java.util.List<Object> messages) {
            return new MessagesLoaded("messages_loaded", messages);
        }
    }

    /** 思考档位同步（off/low/medium/xhigh；实际生效档位，含切换确认回流）。 */
    record ThinkingLevel(String type, String level) implements ChatDelta {
        public static ThinkingLevel of(String level) {
            return new ThinkingLevel("thinking_level", level);
        }
    }

    /** prompt 已受理（对应源 WS 的 response{command:"prompt",success:true}）。 */
    record PromptAccepted(String type) implements ChatDelta {
        public static PromptAccepted of() {
            return new PromptAccepted("prompt_accepted");
        }
    }

    /** 开新思考块。 */
    record ThinkingStart(String type) implements ChatDelta {
        public static ThinkingStart of() {
            return new ThinkingStart("thinking_start");
        }
    }

    /** 思考文本增量。 */
    record ThinkingDelta(String type, String text) implements ChatDelta {
        public static ThinkingDelta of(String text) {
            return new ThinkingDelta("thinking_delta", text);
        }
    }

    /** 正文文本增量。 */
    record TextDelta(String type, String text) implements ChatDelta {
        public static TextDelta of(String text) {
            return new TextDelta("text_delta", text);
        }
    }

    /** 工具调用开始（提前建卡，覆盖「模型构造参数」的反馈空窗；按 id 与 tool_end 配对）。 */
    record ToolStart(String type, String id, String name, Object args) implements ChatDelta {
        public static ToolStart of(String id, String name, Object args) {
            return new ToolStart("tool_start", id, name, args);
        }
    }

    /** 工具调用完整参数回填（toolcall_end 语义）。 */
    record ToolArgs(String type, String id, Object args) implements ChatDelta {
        public static ToolArgs of(String id, Object args) {
            return new ToolArgs("tool_args", id, args);
        }
    }

    /** 工具执行结束。 */
    record ToolEnd(String type, String id, String name, boolean isError, Object result) implements ChatDelta {
        public static ToolEnd of(String id, String name, boolean isError, Object result) {
            return new ToolEnd("tool_end", id, name, isError, result);
        }
    }

    /** 回合终态用量（message_end 语义：total=上下文总 token，stopReason=length 提示截断）。 */
    record MessageUsage(String type, Long total, String stopReason) implements ChatDelta {
        public static MessageUsage of(Long total, String stopReason) {
            return new MessageUsage("message_usage", total, stopReason);
        }
    }

    /**
     * HITL 请求浮层（源 {@code extension_ui_request} 形状：uuid id + method，前端独立
     * ui-requests store 渲染）。M1-6 的 method 恒为 {@code confirm}（危险命令人工确认）。
     */
    record UiRequest(String type, UiRequestPayload request) implements ChatDelta {
        public static UiRequest of(UiRequestPayload request) {
            return new UiRequest("ui_request", request);
        }
    }

    /** 一轮 prompt 处理完成（关闭末尾思考块、置 done、退出 streaming）。 */
    record AgentEnd(String type) implements ChatDelta {
        public static AgentEnd of() {
            return new AgentEnd("agent_end");
        }
    }

    /** 错误（统一复位流式/加载态）。 */
    record Error(String type, String message) implements ChatDelta {
        public static Error of(String message) {
            return new Error("error", message);
        }
    }
}

File: apps/web/src/lib/chat-delta.ts
// chat-delta reducer（纯函数）：ChatDelta → state patch。
// 语义借鉴源 apps/web store/chat-delta.ts（块 id 稳定、末块 append、按 user 边界开新回合），
// 适配本仓后端的 ChatDelta 联合（无 fork/compaction/ready——SSE 断线续传由
// Last-Event-ID + messages_loaded 全量重建解决，无需静默重连状态机）。

import { mapHistory } from "./messages-history";
import {
  newMsgId,
  newTextBlockId,
  newThinkingBlockId,
  type ChatMessage,
  type MessageBlock,
  type ToolCall,
} from "./messages";
import type { ChatDelta } from "./chat-delta-protocol";

/** reducer 可下发的 UI 副作用（ui_request → 独立浮层 store，reducer 保持纯）。 */
export interface DeltaSideEffects {
  /** HITL 确认请求（每条一发，调用方交给 ui-requests store）。 */
  readonly uiRequests: readonly { id: string; method: string; title?: string }[];
  /** 连接错误文本（error 事件）。 */
  readonly errorMessage: string | null;
}

export interface ChatDeltaState {
  readonly messages: readonly ChatMessage[];
  readonly streaming: boolean;
  readonly sessionName: string | null;
  readonly thinkingLevel: string;
  readonly usageTotal?: number;
  readonly stopReason?: string;
}

export const initialChatDeltaState: ChatDeltaState = {
  messages: [],
  streaming: false,
  sessionName: null,
  thinkingLevel: "medium",
};

/** 末尾 assistant 消息（流式目标）；无则 null。 */
function tailAssistant(messages: readonly ChatMessage[]): ChatMessage | null {
  const last = messages[messages.length - 1];
  return last && last.role === "assistant" ? last : null;
}

function withTail(
  messages: readonly ChatMessage[],
  tail: ChatMessage,
): readonly ChatMessage[] {
  return [...messages.slice(0, -1), tail];
}

function ensureTail(state: ChatDeltaState): ChatDeltaState {
  if (tailAssistant(state.messages)) return state;
  return { ...state, messages: [...state.messages, newAssistant()] };
}

function newAssistant(): ChatMessage {
  return { id: newMsgId(), role: "assistant", blocks: [], done: false };
}

/** 关闭末尾思考块（正文/工具 delta 到来时）。 */
function closeThinking(blocks: readonly MessageBlock[]): readonly MessageBlock[] {
  const last = blocks[blocks.length - 1];
  if (last && last.type === "thinking" && last.endedAt === undefined) {
    return [...blocks.slice(0, -1), { ...last, endedAt: Date.now() }];
  }
  return blocks;
}

/**
 * 应用一条 delta，返回新 state 与副作用（一次 set 合并下发）。
 * ui_request / error 不进消息流：前者出副作用，后者由调用方置 error 提示。
 */
export function reduceChatDelta(
  state: ChatDeltaState,
  delta: ChatDelta,
): { state: ChatDeltaState; effects: DeltaSideEffects } {
  const effects: DeltaSideEffects = { uiRequests: [], errorMessage: null };

  const next = (() => {
    switch (delta.type) {
      case "session_info":
        return { ...state, sessionName: delta.sessionName ?? state.sessionName };

      case "thinking_level":
        return { ...state, thinkingLevel: delta.level };

      case "messages_loaded":
        // 全量重建：整体替换（来自 DB 历史投影），退出流式态。
        return {
          ...state,
          messages: mapHistory(delta.messages ?? []),
          streaming: false,
        };

      case "prompt_accepted":
        // 后端无此事件的失败形态（409 在 HTTP 层报），仅幂等确认流式态。
        return state.streaming ? state : { ...state, streaming: true };

      case "thinking_start": {
        const s = ensureTail(state);
        const tail = tailAssistant(s.messages)!;
        return {
          ...s,
          streaming: true,
          messages: withTail(s.messages, {
            ...tail,
            blocks: [
              ...tail.blocks,
              {
                type: "thinking",
                id: newThinkingBlockId(),
                text: "",
                startedAt: Date.now(),
              },
            ],
          }),
        };
      }

      case "thinking_delta": {
        const s = ensureTail(state);
        const tail = tailAssistant(s.messages)!;
        const last = tail.blocks[tail.blocks.length - 1];
        if (last && last.type === "thinking") {
          // append 只换 text 不换 id → React key 稳定
          return {
            ...s,
            streaming: true,
            messages: withTail(s.messages, {
              ...tail,
              blocks: [
                ...tail.blocks.slice(0, -1),
                { ...last, text: last.text + delta.text },
              ],
            }),
          };
        }
        // 无打开的思考块（丢 start 或断线重连续传）：补一块收容
        return {
          ...s,
          streaming: true,
          messages: withTail(s.messages, {
            ...tail,
            blocks: [
              ...tail.blocks,
              {
                type: "thinking",
                id: newThinkingBlockId(),
                text: delta.text,
                startedAt: Date.now(),
              },
            ],
          }),
        };
      }

      case "text_delta": {
        const s = ensureTail(state);
        let tail = tailAssistant(s.messages)!;
        let blocks = closeThinking(tail.blocks);
        const last = blocks[blocks.length - 1];
        if (last && last.type === "text") {
          blocks = [...blocks.slice(0, -1), { ...last, text: last.text + delta.text }];
        } else {
          blocks = [...blocks, { type: "text", id: newTextBlockId(), text: delta.text }];
        }
        tail = { ...tail, blocks };
        return { ...s, streaming: true, messages: withTail(s.messages, tail) };
      }

      case "tool_start": {
        const s = ensureTail(state);
        let tail = tailAssistant(s.messages)!;
        // 已有同 id 卡（断线重连续传重复建卡）：幂等跳过
        if (tail.blocks.some((b) => b.type === "tool" && b.id === delta.id)) {
          return { ...s, streaming: true };
        }
        const call: ToolCall = {
          id: delta.id,
          name: delta.name,
          status: "running",
          args: delta.args,
          startedAt: Date.now(),
        };
        tail = {
          ...tail,
          blocks: [...closeThinking(tail.blocks), { type: "tool", id: delta.id, toolCall: call }],
        };
        return { ...s, streaming: true, messages: withTail(s.messages, tail) };
      }

      case "tool_args": {
        const s = ensureTail(state);
        const tail = tailAssistant(s.messages)!;
        const idx = tail.blocks.findIndex((b) => b.type === "tool" && b.id === delta.id);
        if (idx < 0) return s;
        const b = tail.blocks[idx];
        if (b && b.type === "tool") {
          const blocks = [...tail.blocks];
          blocks[idx] = {
            type: "tool",
            id: delta.id,
            toolCall: { ...b.toolCall, args: delta.args },
          };
          return { ...s, messages: withTail(s.messages, { ...tail, blocks }) };
        }
        return s;
      }

      case "tool_end": {
        const s = ensureTail(state);
        const tail = tailAssistant(s.messages)!;
        const idx = tail.blocks.findIndex((b) => b.type === "tool" && b.id === delta.id);
        if (idx < 0) return s;
        const b = tail.blocks[idx];
        if (b && b.type === "tool") {
          const blocks = [...tail.blocks];
          blocks[idx] = {
            type: "tool",
            id: delta.id,
            toolCall: {
              ...b.toolCall,
              status: delta.isError ? "error" : "done",
              result: delta.result,
              name: delta.name || b.toolCall.name,
              endedAt: Date.now(),
            },
          };
          return { ...s, messages: withTail(s.messages, { ...tail, blocks }) };
        }
        return s;
      }

      case "message_usage":
        return {
          ...state,
          usageTotal: delta.total,
          stopReason: delta.stopReason,
        };

      case "agent_end": {
        // 回合收尾：关思考块、置 done、退出流式。
        const tail = tailAssistant(state.messages);
        if (!tail) return { ...state, streaming: false };
        return {
          ...state,
          streaming: false,
          messages: withTail(state.messages, {
            ...tail,
            done: true,
            blocks: closeThinking(tail.blocks),
          }),
        };
      }

      case "error":
        return { ...state, streaming: false };

      default:
        return state;
    }
  })();

  // ui_request 不改消息流，仅出副作用（上一分支返回原 state）。
  if (delta.type === "ui_request") {
    return {
      state,
      effects: { ...effects, uiRequests: [delta.request] },
    };
  }

  // error 不改消息流，仅出错误提示副作用（调用方置 error 并清 HITL 浮层）。
  if (delta.type === "error") {
    return { state: next, effects: { ...effects, errorMessage: delta.message } };
  }

  return { state: next, effects };
}
